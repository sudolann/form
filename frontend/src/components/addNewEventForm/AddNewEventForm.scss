components$gray: #e4e4e4;

.form {
  max-width: 900px;
  padding-top: 10px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border: solid 1px $gray;
  align-self: center;
  min-width: 600px;
  border-radius: 3px;
  margin-top: 100px;

  .form--item {
    margin-bottom: 20px;
  }

  .form--input {
    background-color: transparent;
    font-size: 1rem;
    padding: 6px 6px 4px 8px;
    border: 0;
    width: 200px;
    color: gray;
    display: inline-block;
    border: 1px solid $gray;
    margin-bottom: 5px;
    border-radius: 3px;

    &:hover,
    &:focus,
    &:active {
      background-color: transparent;
      transition: 0.2s ease;
      outline-color: transparent;
      outline: none;
      border: 1px solid #41a9ff;
    }
  }

  .btn-form {
    margin-bottom: 10px;
  }
}