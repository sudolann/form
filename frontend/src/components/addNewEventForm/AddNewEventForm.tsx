import React, { FunctionComponent, ReactElement, useState } from "react";
import { useForm } from "react-hook-form";
import { ErrorAlert } from "../errorAlert/ErrorAlert";
import { LoadingBar } from "../loadingBar/LoadingBar";
import { useHttpClient } from "../../hooks/useHttpClient";
import { useHistory } from "react-router";
import { Button, DatePicker } from "antd";
import "./AddNewEventForm.scss";

type SubmittedData = { [s: string]: string };


export const AddNewEventForm: FunctionComponent = (): ReactElement => {
  const { register, handleSubmit, errors } = useForm();
  const [date, setDate] = useState<string | undefined>();
  const [newEventId, setNewEventId] = useState<string | null>(null);
  const history = useHistory();
  const { isLoading, error, sendRequest, setError } = useHttpClient();

  const addNewNewEvent = async (
    name: string,
    email: string,
    date: string
  ): Promise<void> => {
    try {
      const responseData = await sendRequest(
        `http://localhost:${process.env.REACT_APP_BACKEND_PORT}/addNewEvent`,
        "POST",
        JSON.stringify({
          name,
          email,
          date,
        }),
        {
          "Content-Type": "application/json",
        }
      );
      setNewEventId(responseData.eventId);
    } catch (err) {
      setError(err.message);
    }
  };
  const onSubmit = (data: SubmittedData): void => {
    const { name, email } = data;
    date && addNewNewEvent(name, email, date)
  };

  if (newEventId) {
    history.push({ pathname: `/event/${newEventId}` });
  }
  


  if (isLoading) {
    return <LoadingBar />;
  }


  return (
    <form onSubmit={handleSubmit(onSubmit)} className="form">
      <h2>Add New Event</h2>
      {error && <ErrorAlert errorMessage={error}/>}
      <div className="form--item">
        <input
          type="text"
          name="name"
          placeholder="Name"
          ref={register({ required: "Name is required", minLength: 3 })}
          className="form--input"
        />
        {errors.name && <ErrorAlert errorMessage="Name is required" />}
      </div>
      <div className="form--item">
        <input
          type="text"
          name="email"
          placeholder="Email"
          ref={register({
            required: "Email is required",
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: "invalid email address"
            }
          })}
          className="form--input"
        />
        {errors.email && <ErrorAlert errorMessage={errors.email.message} />}
      </div>

      <div className="form--item">
        <DatePicker
            name="startDate"
            onChange={(_date: any, dateString: string):void=> setDate(dateString)}
            style={{width: '200px'}}
            size="large"
        />
        {Object.values(errors).length > 0 && !date && <ErrorAlert errorMessage="Date is required" />}

      </div>

    <Button type='primary' htmlType='submit' className="btn-form">
        Add event
    </Button>
    </form>
  );
};
