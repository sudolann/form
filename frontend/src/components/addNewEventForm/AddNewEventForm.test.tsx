import React from 'react';
import { AddNewEventForm } from './AddNewEventForm';
import '@testing-library/jest-dom/extend-expect';

import { render, fireEvent, screen, getAllByTestId, MatcherFunction } from '@testing-library/react';

const setup = (labelText: string) => {
  const utils = render(<AddNewEventForm />);
  const input = utils.getByLabelText(labelText);
  return {
    input,
    ...utils,
  };
};

describe('Inputs validation', (): void => {
  it.each`
    inputName        | attr
    ${'Name'}        | ${['type', 'text']}
    ${'Email'}       | ${['type', 'email']}
    ${'Select date'} | ${['name', 'event date']}
  `(
    '$inputName has an attribute $attr[0] equal $attr[1] and their value is empty by default',
    ({ inputName, attr }): void => {
      const { getByPlaceholderText } = render(<AddNewEventForm />);
      expect(getByPlaceholderText(inputName)).toHaveAttribute(attr[0], attr[1]);
      expect(getByPlaceholderText(inputName)).toHaveAttribute('value', '');
    },
  );
  describe('renders validation message when added invalid input value on', () => {
    it('Name input', async () => {
      const { findByText } = render(<AddNewEventForm />);
      const { input } = setup('Event Name');
      fireEvent.change(input, { target: { value: 'ab' } });

      const alert = await findByText(/Event name must be 3 characters long!/i, { exact: false });

      expect(alert).toBeInTheDocument();
    });
    it('Email input', async () => {
      const { findByText } = render(<AddNewEventForm />);
      const { input } = setup('Email Address');
      fireEvent.change(input, { target: { value: 'te' } });

      const alert = await findByText(/Email is not valid!/i, { exact: false });

      expect(alert).toBeInTheDocument();
    });
  });
});

describe('AddNewEventForm', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<AddNewEventForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('displays title', () => {
    const { getByText } = render(<AddNewEventForm />);
    const title = getByText(/Add New Event/i);
    expect(title).toBeInTheDocument();
  });
  it('renders disabled button by default', () => {
    const { getByRole } = render(<AddNewEventForm />);
    expect(getByRole('button')).toBeDisabled();
  });
});
