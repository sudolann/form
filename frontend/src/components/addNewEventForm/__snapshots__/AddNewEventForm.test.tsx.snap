// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewEventForm should take a snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <h2>
      Add New Event
    </h2>
    <div
      class="form--item"
    >
      <label
        for="name"
      >
         Event Name
      </label>
      <input
        class="form--input"
        id="name"
        name="name"
        placeholder="Name"
        type="text"
        value="Event name must be 3 characters long!"
      />
      <p
        class="alert"
        data-testid="error"
      >
        Event name must be 3 characters long!
      </p>
    </div>
    <div
      class="form--item"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form--input"
        id="email"
        name="email"
        placeholder="Email"
        type="email"
        value="Email is not valid!"
      />
      <p
        class="alert"
        data-testid="error"
      >
        Email is not valid!
      </p>
    </div>
    <div
      class="form--item"
    >
      <div
        class="ant-picker ant-picker-large"
        style="width: 200px;"
      >
        <div
          class="ant-picker-input"
        >
          <input
            autocomplete="off"
            name="event date"
            placeholder="Select date"
            readonly=""
            size="12"
            title=""
            value=""
          />
          <span
            class="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              class="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
      <p
        class="alert"
        data-testid="error"
      />
    </div>
    <button
      class="ant-btn btn-form ant-btn-primary"
      disabled=""
      type="submit"
    >
      <span>
        Add event
      </span>
    </button>
  </form>
</DocumentFragment>
`;
